### PR TITLE
[BC5] Add/use productPlanIdentifier to EntitlementInfo and CustomerInfo

### PR DESCRIPTION
**Pull request by joshdholtz**
### Motivation

[CF-1158](https://revenuecats.atlassian.net/browse/CF-1158)

`product_plan_identifier` has been added to entitlements and subscriptions in the full customer response. Developers should be able to see what base plan was purchased in the `EntitlementInfo`.

### Description

- Added `productPlanIdentifier` to `EntitlementInfo`
- `allExpirationDatesByProduct` and `allPurchaseDatesByProduct` on `CustomerInfo` are now keyed by `subId:basePlanId`
  - The decision to key the maps like this was because `storeProduct.id` is also in the `subId:basePlanId` format
- Added tests for `allExpirationDatesByProduct` and `allPurchaseDatesByProduct`

[CF-1158]: https://revenuecats.atlassian.net/browse/CF-1158?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: common/src/main/java/com/revenuecat/purchases/common/CustomerInfoFactories.kt [hunk 1]
```diff
@@ -65,3 +65,42 @@ fun JSONObject.buildCustomerInfo(): CustomerInfo {
...
+ * @throws [JSONException] If the json is invalid.
+ */
+private fun JSONObject.parseDates(jsonKey: String): HashMap<String, Date?> {
+    val expirationDates = HashMap<String, Date?>()
+
+    val it = keys()
+    while (it.hasNext()) {
+        val productId = it.next()
+        val basePlanId = getJSONObject(productId).optString("product_plan_identifier")
+            .takeIf { it.isNotEmpty() }?.split(":")?.firstOrNull()
```
Comment (swehner) [inline common/src/main/java/com/revenuecat/purchases/common/CustomerInfoFactories.kt:97]: Why the split by ":" here? The plan identifier should just be basePlanId without anything else, right?
https://github.com/RevenueCat/khepri/blob/d1aec0d684e21a682b55234447711fcd4d8c5f82/khepri/serializers/subscriber_serializer.py#L243

File: common/src/main/java/com/revenuecat/purchases/common/CustomerInfoFactories.kt [hunk 2]
```diff
@@ -65,3 +65,42 @@ fun JSONObject.buildCustomerInfo(): CustomerInfo {
...
+
+        val expirationObject = getJSONObject(productId)
+
+        val key = basePlanId?.let { "$productId:$it" } ?: productId
```
Comment (swehner) [inline common/src/main/java/com/revenuecat/purchases/common/CustomerInfoFactories.kt:101]: Why this concatenation here? We don't really want to force the product identifiers to be `ProductId:BasePlanId` necessarily.

File: common/src/main/java/com/revenuecat/purchases/common/CustomerInfoFactories.kt [hunk 3]
```diff
@@ -65,3 +65,42 @@ fun JSONObject.buildCustomerInfo(): CustomerInfo {
...
+private fun JSONObject.parsePurchaseDates(): Map<String, Date?> {
+    return parseDates("purchase_date")
+}
+
+/**
+ * Parses dates that match a JSON key in a JSONObject
+ * @param jsonKey Key of the dates to deserialize from the JSONObject
+ * @throws [JSONException] If the json is invalid.
+ */
+private fun JSONObject.parseDates(jsonKey: String): HashMap<String, Date?> {
```
Comment (swehner) [inline common/src/main/java/com/revenuecat/purchases/common/CustomerInfoFactories.kt:90]: Is `jsonKey` used at all here?

File: common/src/main/java/com/revenuecat/purchases/common/CustomerInfoFactories.kt (hunk 1 above)
Comment (joshdholtz) [reply common/src/main/java/com/revenuecat/purchases/common/CustomerInfoFactories.kt:97]: Oh, is it?! I assumed the offer also was present in it but I'm not sure why 🤦‍♂️

I will remove that then!

File: public/src/main/java/com/revenuecat/purchases/EntitlementInfo.kt [hunk 4]
```diff
@@ -23,6 +23,7 @@ import java.util.Date
  * @property store The store where this entitlement was unlocked from. Either: APP_STORE,
  * MAC_APP_STORE, PLAY_STORE, STRIPE, PROMOTIONAL or UNKNOWN_STORE.
  * @property productIdentifier The product identifier that unlocked this entitlement.
+ * @property productPlanIdentifier The product plan identifier that unlocked this entitlement.
```
Comment (beylmk) [inline public/src/main/java/com/revenuecat/purchases/EntitlementInfo.kt:26]: maybe we can call out that this is google-only, and what it is for google?

**Review by beylmk — APPROVED** [review]
🚢 it!

**Pull request merged**